packages/loopring_v3/circuit/Gadgets/MathGadgets.h

```diff
@@ -67,15 +123,106 @@ class Constants : public GadgetT
         pb.add_r1cs_constraint(ConstraintT(_100000, FieldT::one(), ethsnarks::FieldT(100000)), "._100000");
+        pb.add_r1cs_constraint(ConstraintT(indexBase, FieldT::one(), ethsnarks::FieldT(INDEX_BASE)), ".indexBase");
         pb.add_r1cs_constraint(ConstraintT(emptyTradeHistory, FieldT::one(), ethsnarks::FieldT(EMPTY_TRADE_HISTORY)), ".emptyTradeHistory");
         pb.add_r1cs_constraint(ConstraintT(maxAmount, FieldT::one(), ethsnarks::FieldT(MAX_AMOUNT)), ".maxAmount");
         pb.add_r1cs_constraint(ConstraintT(maxConcurrentOrderIDs, FieldT::one(), ethsnarks::FieldT(MAX_CONCURRENT_ORDERIDS)), ".maxConcurrentOrderIDs");
+        pb.add_r1cs_constraint(ConstraintT(dummyPublicKeyX, FieldT::one(), ethsnarks::FieldT("132404916167441185773716937639098950030214269269071041759116060313694190797")), ".dummyPublicKeyX");
+        pb.add_r1cs_constraint(ConstraintT(dummyPublicKeyY, FieldT::one(), ethsnarks::FieldT("6933274320914065805670637410453081675154127044926882796951068647148079547843")), ".dummyPublicKeyY");
+        pb.add_r1cs_constraint(ConstraintT(txTypeTransfer, FieldT::one(), ethsnarks::FieldT(int(TransactionType::Transfer))), ".txTypeTransfer");
+        pb.add_r1cs_constraint(ConstraintT(txTypeNewAccount, FieldT::one(), ethsnarks::FieldT(int(TransactionType::AccountNew))), ".txTypeNewAccount");
+
+        pb.add_r1cs_constraint(ConstraintT(pow10_c0, FieldT::one(), ethsnarks::FieldT(POW10_C0)), ".pow10_c0");
+        pb.add_r1cs_constraint(ConstraintT(pow10_c1, FieldT::one(), ethsnarks::FieldT(POW10_C1)), ".pow10_c1");
+        pb.add_r1cs_constraint(ConstraintT(pow10_c2, FieldT::one(), ethsnarks::FieldT(POW10_C2)), ".pow10_c2");
+        pb.add_r1cs_constraint(ConstraintT(pow10_c3, FieldT::one(), ethsnarks::FieldT(POW10_C3)), ".pow10_c3");
+    }
+};
+
+class DualVariableGadget : public libsnark::dual_variable_gadget<FieldT>
+{
+public:
+    bool fromPacked = false;
+    bool fromBits = false;
+
+    DualVariableGadget(
+        ProtoboardT& pb,
+        const size_t width,
+        const std::string& prefix
+    ) :
+        libsnark::dual_variable_gadget<FieldT>(pb, width, prefix)
+    {
+
+    }
+
+    DualVariableGadget(
+        ProtoboardT& pb,
+        const VariableT& value,
+        const size_t width,
+        const std::string& prefix
+    ) :
+        libsnark::dual_variable_gadget<FieldT>(pb, value, width, prefix)
+    {
+        fromPacked = true;
+    }
+
+    DualVariableGadget(
+        ProtoboardT& pb,
+        const VariableArrayT& bits,
+        const std::string& prefix
+    ) :
+        libsnark::dual_variable_gadget<FieldT>(pb, bits, prefix)
+    {
+        fromBits = true;
+    }
+
+    void generate_r1cs_witness()
+    {
+        if (fromPacked)
+        {
+            generate_r1cs_witness_from_packed();
+        }
+        if (fromBits)
+        {
+            generate_r1cs_witness_from_bits();
+        }
+    }
+
+    void generate_r1cs_witness(ProtoboardT& pb, const FieldT& value)
```

Comment:
Both should be `assert(!fromPacked && !fromBits)` because otherwise the value is deduced from the input variable passed into the constructor. Will add this check.

Important to note here is that these asserts don't actually impact the circuit itself at all, but are still useful to enforce correct use of the gadgets which could find potential issues when misused (though the witness generation itself is also unrelated to the security of the circuits, only the constraints themselves matter).